:sparkles:Add NSX-VPC Network Provider Support

silvery1622 opened the pull request:
This commit introduces NSX-T Virtual Private Cloud (nsx-vpc) as a new network provider for the vSphere Supervisor Cluster in anticipation of the upcoming vSphere 9 release. Key Changes:
    nsx-vpc implemented as a new NetworkProvider using nsx-operator libs.
    Skipped VM Readiness check as nsx-vpc offers private network access.
    Added unit tests for nsx-vpc network provider.

**What this PR does / why we need it**:
The vSphere Supervisor Cluster currently accommodates two network types: vSphere Distributed Switch (VDS) and NSX-T network. With the impending release of vSphere 9, a third network type will be introduced: NSX-T Virtual Private Cloud (nsx-vpc). 
To adapt this change,  CAPV need to add support for nsx-vpc  as a new network provider.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2847

zhanggbj commented:
/ok-to-test

laozc commented on pkg/services/interfaces.go:
```diff
@@ -71,6 +71,9 @@ type NetworkProvider interface {
 	HasLoadBalancer() bool
 
+	// SupportVMReadinessProbe indicates whether this provider support vm readiness probe.
+	SupportVMReadinessProbe() bool
```
I think `SupportsVMRedinessProbe` or `VMReadinessProbeSupported` would be more aligned with general naming convention.

laozc commented on pkg/manager/manager.go:
```diff
@@ -22,6 +22,7 @@ import (
 
 	"github.com/pkg/errors"
 	netopv1 "github.com/vmware-tanzu/net-operator-api/api/v1alpha1"
+	nsxoperatorv1 "github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
```
May shorten to nsxopv1 as net-operator also gets shortened.

laozc commented on pkg/services/network/nsxvpc_provider.go:
```diff
@@ -0,0 +1,178 @@
+	return true
+}
+
+// nsxvpcNetworkProvider doesn't support VM readiness probe
```
I believe the comments here belongs to the docstring of the method.
You may move it down to the function body or cite these lines as Note:

silvery1622 commented:
/retest

laozc commented on pkg/services/network/nsxvpc_provider.go:
```diff
@@ -0,0 +1,178 @@
+	clusterName := ctx.VSphereCluster.Name
+	namespace := ctx.VSphereCluster.Namespace
+	for _, condition := range subnetset.Status.Conditions {
+		if condition.Type == "Ready" && condition.Status != "True" {
```
Are there any constants we can refer to?

zhanggbj commented on pkg/services/network/network_test.go:
```diff
@@ -470,7 +501,124 @@ var _ = Describe("Network provider", func() {
+				Expect(createdSubnetSet.Spec).To(Equal(initialSubnetSetSpec), "SubnetSetSpec should not have been modified")
+			})
+
+			It("GetVMServiceAnnotations", func() {
```
Better to make the It description more  human-readable

zhanggbj reviewed (commented):
Thanks for the PR! Just left some nits :-)

I think it would be great to add some test output from your local setup in the PR comments for reference, as upstream doesn't have such infra to test it in e2e or integration test.

silvery1622 commented:
Thanks for @zhanggbj 's comments.
I have tested the changes in a VCF env with nsx-vpc enabled. Attached the output of cluster/vspherecluster/vspheremachine/virtualmachine and capv log.
[capv.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14739792/capv.log)
[cluster.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14739813/cluster.log)
[vspherecluster.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14739814/vspherecluster.log)
[vspheremachine.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14739815/vspheremachine.log)
[virtualmachine.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14739817/virtualmachine.log)
[subnetset.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14739818/subnetset.log)
[subnetport.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14739819/subnetport.log)

zhanggbj commented on pkg/services/network/network_test.go:
```diff
@@ -195,13 +196,42 @@ var _ = Describe("Network provider", func() {
+			AfterEach(func() {
+				Expect(err).ToNot(HaveOccurred())
+				Expect(vm.Spec.NetworkInterfaces[0].NetworkName).To(Equal(GetNsxVpcSubnetSetName(vSphereCluster.Name)))
+				Expect(vm.Spec.NetworkInterfaces[0].NetworkType).To(Equal("nsx-t-subnetset"))
```
same as above

silvery1622 replied:
Modified the description.

zhanggbj commented on pkg/services/network/network_test.go:
```diff
@@ -470,7 +501,124 @@ var _ = Describe("Network provider", func() {
+			})
+		})
+
+		Context("with nsxvpc network provider and subnetset exists", func() {
```
```suggestion
		Context("with nsx-vpc network provider and subnetset exists", func() {
```

zhanggbj reviewed (changes requested):
Generally it looks good :-) Just added some other nits in the second round review.

silvery1622 commented on pkg/services/network/constants.go:
```diff
@@ -22,6 +22,10 @@ const (
 	// NSXTVNetSelectorKey is also defined in VM Operator.
 	NSXTVNetSelectorKey = "ncp.vmware.com/virtual-network-name"
+	// NSXVPCTypeNetwork is the name of the NSX VPC network type.
+	NSXVPCTypeNetwork = "nsx-t-subnetset"
```
"nsx-vpc" is the network provider type passed by supervisor cluster while "nsx-t-subnetset" is the valid network type defined by vm-operator.

silvery1622 commented on pkg/services/network/network_test.go:
```diff
@@ -195,13 +196,42 @@ var _ = Describe("Network provider", func() {
+				err = np.ConfigureVirtualMachine(ctx, clusterCtx, vm)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(vm.Spec.NetworkInterfaces).To(HaveLen(1))
+				Expect(vm.Spec.NetworkInterfaces[0].NetworkType).To(Equal("nsx-t-subnetset"))
```
Modified the description.

chrischdi commented on pkg/services/interfaces.go:
```diff
@@ -71,6 +71,9 @@ type NetworkProvider interface {
 	HasLoadBalancer() bool
 
+	// SupportsVMRedinessProbe indicates whether this provider support vm readiness probe.
+	SupportsVMRedinessProbe() bool
```
```suggestion
	SupportsVMReadinessProbe() bool
```

chrischdi replied:
Isn't this more a `RequiresVMReadinessProbe`?

silvery1622 replied:
If we're using Avi as a Load Balancer provider, we do not require the VM readiness probe because Avi already has an in-built service endpoint health checker. However, NSX LoadBalancer doesn't share this feature. Hence, CAPV introduced the readiness probe.
The Kubernetes 1.19 update has an added feature, which allows retries for the kubeadm join operation. Ideally, if these retries can be successful before the timeout period (5 minutes), we can potentially eliminate the VM readiness check for all network provider types. However, we'd need to run a scale test before we entirely remove the VM readiness check for vsphere-vds and nsxt.
Concerning nsx-vpc, the VM readiness check would be unsuccessful, because the node network remains unreachable from the supervisor master node. That's the reason why I suggest we name the function to SupportsVMReadinessProbe().

chrischdi commented on pkg/services/network/constants.go:
```diff
@@ -22,6 +22,10 @@ const (
 	// NSXTVNetSelectorKey is also defined in VM Operator.
 	NSXTVNetSelectorKey = "ncp.vmware.com/virtual-network-name"
+	// NSXVPCTypeNetwork is the name of the NSX VPC network type.
+	NSXVPCTypeNetwork = "nsx-t-subnetset"
```
We should either rename the value or the key to match.

So either `NSXTSubnetSetTypeNetwork` or `NSXTVPCTypeNetwork = "nsx-t-vpc"`

chrischdi commented on pkg/services/network/constants.go:
```diff
@@ -22,6 +22,10 @@ const (
+	// NSXVPCTypeNetwork is the name of the NSX VPC network type.
+	NSXVPCTypeNetwork = "nsx-t-subnetset"
+	// NSXVPCSubnetSetSelectorKey has to be aligned with VM Operator design.
+	NSXVPCSubnetSetSelectorKey = "nsx-operator.vmware.com/subnetset-name"
```
Is this some key coming from the nsx-operator? Can we add a reference link to it if so?
What is it used for?

We should align the naming to the decision above if possible.

silvery1622 replied:
These are the annotation keys designed for the VirtualMachineService and LoadBalancer type service, originally defined in CAPW. It's quite uncertain if vm-operator/ncp/nsx-operator utilize them anymore. We should contemplate removing all of them once their usage in vm-operator/ncp/nsx-operator has been confirmed and the adjustment could be undertaken in a separate, dedicated fix.

chrischdi commented on controllers/vspheremachine_controller.go:
```diff
@@ -90,12 +90,12 @@ func AddMachineControllerToManager(ctx context.Context, controllerManagerContext
 		if err != nil {
 			return errors.Wrap(err, "failed to create a network provider")
 		}
 		r.networkProvider = networkProvider
```
Optional: Do we need to duplicate r.networkProvider 🤔

silvery1622 replied:
The r.networkProvider is utilized within the setVMModifiers function, where it invokes the ConfigureVirtualMachine interface of the network provider.

chrischdi commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -38,12 +38,14 @@ import (
 type VmopMachineService struct {
-	Client client.Client
+	Client          client.Client
+	NetworkProvider services.NetworkProvider
```
Wondering if we should have a bool here instead of a full blown network provider.

silvery1622 replied:
That's a nice idea.

chrischdi commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -345,7 +347,15 @@ func (v *VmopMachineService) reconcileVMOperatorVM(ctx context.Context, supervis
 		// that is not yet ready.
-		if infrautilv1.IsControlPlaneMachine(supervisorMachineCtx.Machine) && supervisorMachineCtx.Cluster.Status.ControlPlaneReady {
+		if v.NetworkProvider == nil {
+			return errors.Errorf(
+				"missing network provider for %s %s/%s",
+				supervisorMachineCtx.Machine.GroupVersionKind(),
+				supervisorMachineCtx.Machine.Namespace,
+				supervisorMachineCtx.Machine.Name)
```
If we hit this case, then it is not only missing for this object, but for all instead.

silvery1622 replied:
With the change to update the NetworkProvider in VmopMachineService to a boolean, this particular check has become unnecessary.

chrischdi commented on pkg/services/network/nsxvpc_provider.go:
```diff
@@ -0,0 +1,192 @@
+// verifyNsxVpcSubnetSetStatus checks the status conditions of a given SubnetSet within a cluster context.
+// If the subnet isn't ready, it is marked as false, and the function returns an error.
+// If the subnet is ready, the function updates the VSphereCluster with a "true" status and returns nil.
+func (vp *nsxvpcNetworkProvider) verifyNsxVpcSubnetSetStatus(ctx *vmware.ClusterContext, subnetset *nsxopv1.SubnetSet) error {
```
I'd prefer not having `vmware.ClusterContext` as parameter here (I think only VSphereCluster is used?), please pass down instead what really is requred.

also not name it ctx, `ctx` should only be used for real `context.Context` objects to not confuse when reading code :-)

silvery1622 replied:
Will address it.

chrischdi commented on pkg/services/network/constants.go:
```diff
@@ -22,6 +22,11 @@ const (
 	// NSXTVNetSelectorKey is also defined in VM Operator.
 	NSXTVNetSelectorKey = "ncp.vmware.com/virtual-network-name"
+	// NSXTSubnetSetTypeNetwork is the name of the NSX VPC network type.
+	NSXTSubnetSetTypeNetwork = "nsx-t-subnetset"
```
```suggestion
	NSXTVPCSubnetSetNetworkType = "nsx-t-subnetset"
```

chrischdi commented on pkg/services/network/constants.go:
```diff
@@ -22,6 +22,11 @@ const (
+	NSXTSubnetSetTypeNetwork = "nsx-t-subnetset"
+	// NSXVpcSubnetSetSelectorKey has to be aligned with VM Operator design. Please refer to:
+	// https://github.com/vmware-tanzu/vm-operator/blob/main/api/v1alpha1/virtualmachine_types.go#L149.
+	NSXVpcSubnetSetSelectorKey = "nsx-operator.vmware.com/subnetset-name"
```
```suggestion
	NSXTVPCSubnetSetSelectorKey = "nsx-operator.vmware.com/subnetset-name"
```

chrischdi commented on pkg/services/network/constants.go:
```diff
@@ -22,6 +22,11 @@ const (
+	// NSXTSubnetSetTypeNetwork is the name of the NSX VPC network type.
+	NSXTSubnetSetTypeNetwork = "nsx-t-subnetset"
+	// NSXVpcSubnetSetSelectorKey has to be aligned with VM Operator design. Please refer to:
+	// https://github.com/vmware-tanzu/vm-operator/blob/main/api/v1alpha1/virtualmachine_types.go#L149.
```
This reference is for the above variable, not for this one right?

chrischdi commented on pkg/services/network/nsxvpc_provider.go:
```diff
@@ -0,0 +1,187 @@
+/*
```
let's rename this file to nsxt_vpc_provider.go

chrischdi commented on pkg/services/network/nsxvpc_provider.go:
```diff
@@ -0,0 +1,187 @@
+)
+
+// nsxvpcNetworkProvider provisions nsx-vpc type cluster network.
+type nsxvpcNetworkProvider struct {
```
```suggestion
type nsxtVPCNetworkProvider struct {
```

chrischdi replied:
Also rename all relevant funcs to nsxt...

chrischdi commented on pkg/services/vmoperator/vmopmachine.go:
```diff
@@ -43,7 +43,8 @@ import (
 type VmopMachineService struct {
-	Client client.Client
+	Client                   client.Client
+	PerformsVMReadinessProbe bool
```
```suggestion
	ConfigureControlPlaneVMReadinessProbe bool
```

chrischdi commented on pkg/services/network/nsxvpc_provider.go:
```diff
@@ -0,0 +1,187 @@
+}
+
+// NsxVpcNetworkProvider returns an instance of nsx-vpc type network provider.
+func NsxVpcNetworkProvider(client client.Client) services.NetworkProvider {
```
```suggestion
func NSXTVpcNetworkProvider(client client.Client) services.NetworkProvider {
```

silvery1622 commented:
/retest

sbueringer commented:
Apologies for the delay. I'll try to review this soon

chrischdi reviewed (commented):
Just did see that I forgot to send out my review...

sbueringer commented on pkg/services/network/nsxt_vpc_provider.go:
```diff
@@ -0,0 +1,187 @@
+func (vp *nsxtVPCNetworkProvider) verifyNsxtVpcSubnetSetStatus(vspherecluster *vmwarev1.VSphereCluster, subnetset *nsxopv1.SubnetSet) error {
+	clusterName := vspherecluster.Name
+	namespace := vspherecluster.Namespace
+	for _, condition := range subnetset.Status.Conditions {
```
It looks like we assume the network is ready when no ready condition is set. This looks like a race condition to me (in the case where the other controller is not setting ready: false quickly enough)

As far as I can tell we have the same problem in (at least) the nsxt_provider

silvery1622 replied:
Yeah, I get your point. How about I addressed the issue for nsx-vpc in this pull request and raise another pull request for fix in nsxt part.

sbueringer commented on pkg/services/network/nsxt_vpc_provider.go:
```diff
@@ -0,0 +1,187 @@
+func (vp *nsxtVPCNetworkProvider) ProvisionClusterNetwork(ctx context.Context, clusterCtx *vmware.ClusterContext) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	cluster := clusterCtx.VSphereCluster
```
Let's do this instead please:
```go
	cluster := clusterCtx.VSphereCluster
	networkNamespace := cluster.Namespace
	networkName := cluster.Name
	log = log.WithValues("SubnetSet", klog.KRef(networkNamespace, networkName))

	log.Info("Provisioning ")
	defer log.Info("Finished provisioning")

	subnetset := &nsxopv1.SubnetSet{
		ObjectMeta: metav1.ObjectMeta{
			Namespace: networkNamespace,
			Name:      networkName,
		},
		Spec: nsxopv1.SubnetSetSpec{
			AdvancedConfig: nsxopv1.AdvancedConfig{
				StaticIPAllocation: nsxopv1.StaticIPAllocation{
					Enable: true,
				},
			},
		},
	}
```
(uses a proper k/v pair for logging, maps both namespace and name to network{Name,Namespace} not just one of them)

silvery1622 replied:
Thanks for nice suggestion. Will addressed nsx-vpc part here and fix nsxt part in a separate pull request.

sbueringer commented on pkg/services/network/nsxt_vpc_provider.go:
```diff
@@ -0,0 +1,187 @@
+			subnetset,
+			vp.client.Scheme(),
+		); err != nil {
+			return errors.Wrapf(
+				err,
+				"error setting %s/%s as owner of %s/%s",
+				clusterCtx.VSphereCluster.Namespace,
+				clusterCtx.VSphereCluster.Name,
+				subnetset.Namespace,
+				subnetset.Name,
+			)
```
```suggestion
			return errors.Wrapf(err,"error setting %s as owner of %s", klog.KObj(clusterCtx.VSphereCluster), klog.KObj(subnetset))
```
(let's make a similar change to nsxt_provider.go)

sbueringer commented on pkg/services/network/constants.go:
```diff
@@ -22,6 +22,11 @@ const (
 	NSXTVNetSelectorKey = "ncp.vmware.com/virtual-network-name"
+	// NSXTVPCSubnetSetNetworkType is the name of the NSX VPC network type. Please refer to:
+	// https://github.com/vmware-tanzu/vm-operator/blob/main/api/v1alpha1/virtualmachine_types.go#L149.
+	NSXTVPCSubnetSetNetworkType = "nsx-t-subnetset"
```
Just to double check. This is the correct value? I saw "nsx-t-subnet" used in some document.

silvery1622 replied:
Yes, "nsx-t-subnet" and "nsx-t-subnetset" are 2 similar network types and "nsx-t-subnetset"  has the ability of autoscale. We will only use "nsx-t-subnetset" for cluster network.

sbueringer commented on pkg/services/network/network_test.go:
```diff
@@ -470,7 +500,124 @@ var _ = Describe("Network provider", func() {
+				client = fake.NewClientBuilder().WithScheme(scheme).WithRuntimeObjects(configmapObj, systemNamespaceObj).Build()
+				vpcNp, _ = NSXTVpcNetworkProvider(client).(*nsxtVPCNetworkProvider)
+				np = vpcNp
+				err = np.ProvisionClusterNetwork(ctx, clusterCtx)
```
Should we check that this error hasn't occured?

sbueringer replied:
Please also check in other cases below

silvery1622 replied:
Yes, we should check. But we need to add a mock version of ProvisionClusterNetwork to ignore the err regards subnetset ready condition for there is no net provider controller in UT to set it.

sbueringer commented on pkg/services/network/network_test.go:
```diff
@@ -470,7 +500,124 @@ var _ = Describe("Network provider", func() {
+
+				// Presumably there's code here that might modify the SubnetSet...
+				Expect(err).ToNot(HaveOccurred())
+				subnetset, localerr := np.GetClusterNetworkName(ctx, clusterCtx)
```
Why not just call this err?

sbueringer replied:
Same in other cases below

sbueringer replied:
I see, this is used in other cases above as well. But I don't get the point why we:
a) don't check errors that occur in BeforeEach directly in BeforeEach
b) why we store errors in localerr. It might make sense if we wanted to preserve the value of err, but we usually just overwrite it a bit below

Let's please get rid of all these localerr's and let's move the err checks up.

(fine to do it for existing code in a follow-up PR, but let's adjust the new code we add here in this PR)

sbueringer commented on pkg/services/network/network_test.go:
```diff
@@ -470,7 +500,124 @@ var _ = Describe("Network provider", func() {
+				initialSubnetSetSpec = initialSubnetSet.Spec
+
+				// Presumably there's code here that might modify the SubnetSet...
+				Expect(err).ToNot(HaveOccurred())
```
```suggestion
```
err was checked above already

sbueringer replied:
I think this comment got lost. We still have the duplicate err check

silvery1622 replied:
Dropped the dup check now.

sbueringer commented on pkg/services/network/network_test.go:
```diff
@@ -470,7 +500,124 @@ var _ = Describe("Network provider", func() {
+			})
+
+			It("should not update subnetset", func() {
+				var initialSubnetSetSpec nsxopv1.SubnetSetSpec
```
```suggestion
```

sbueringer commented on pkg/services/network/network_test.go:
```diff
@@ -470,7 +500,124 @@ var _ = Describe("Network provider", func() {
+					Namespace: dummyNs,
+				}, initialSubnetSet)
+				Expect(err).NotTo(HaveOccurred())
+				initialSubnetSetSpec = initialSubnetSet.Spec
```
```suggestion
```

sbueringer commented on pkg/services/network/network_test.go:
```diff
@@ -470,7 +500,124 @@ var _ = Describe("Network provider", func() {
+
+				Expect(err).ToNot(HaveOccurred())
+				// Check that the SubnetSetSpec was not changed
+				Expect(createdSubnetSet.Spec).To(Equal(initialSubnetSetSpec), "SubnetSetSpec should not have been modified")
```
```suggestion
				Expect(createdSubnetSet.Spec).To(Equal(initialSubnetSet.Spec), "SubnetSetSpec should not have been modified")
```

sbueringer commented on pkg/services/network/network_test.go:
```diff
@@ -470,7 +500,124 @@ var _ = Describe("Network provider", func() {
+			})
+
+			It("should create subnetset with new spec", func() {
+				Expect(err).ToNot(HaveOccurred())
```
Let's please move this up in l.561

sbueringer replied:
I think this one got lost. We should error check after l.571

silvery1622 replied:
Thanks for reminder.

sbueringer reviewed (commented):
Overall looks good. Mostly just nits.

If you have some time to work on it, we can try to get this done for the upcoming v1.10.0 release (feel free to ping me for quick follow-up reviews)

sbueringer commented:
@silvery1622 ^^

sbueringer commented on pkg/services/vmoperator/vmopmachine_test.go:
```diff
@@ -614,7 +615,7 @@ var _ = Describe("GetMachinesInCluster", func() {
 
 	controllerManagerContext := fake.NewControllerManagerContext(initObjs...)
-	vmService := VmopMachineService{Client: controllerManagerContext.Client}
+	vmService := VmopMachineService{Client: controllerManagerContext.Client, ConfigureControlPlaneVMReadinessProbe: network.DummyLBNetworkProvider().HasLoadBalancer()}
```
Nice catch!

chrischdi commented on pkg/manager/network.go:
```diff
@@ -27,6 +27,8 @@ import (
 )
 
 const (
+	// NSXVPCNetworkProvider identifies the nsx-vpc network provider.
+	NSXVPCNetworkProvider = "nsx-vpc"
```
also fine for me :-)

silvery1622 replied:
How about NSX-VPC?  NSX-V is not in the support scope of vSphere with TKG and NSX won't introduce ambiguity here.  The original idea of using nsx-vpc is to be consistent with the networkprovider name in supervisorproperties of Supervisor cluster. But we definitely could do the covert when generating CAPV manifest(Require a change in supervisor service pakage).

sbueringer replied:
I'm fine with not adding the `t`. How problematic is it to change the supervisor service package? Do we also use upper case NSX there?

sbueringer commented on pkg/manager/manager.go:
```diff
@@ -22,6 +22,7 @@ import (
 
 	"github.com/pkg/errors"
 	netopv1 "github.com/vmware-tanzu/net-operator-api/api/v1alpha1"
+	nsxopv1 "github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
```
Interesting, thx! I didn't actually check the v1alpha2 package, just saw it exists :)

silvery1622 commented on pkg/services/network/constants.go:
```diff
@@ -22,6 +22,11 @@ const (
+	// NSXTVPCSubnetSetNetworkType is the name of the NSX VPC network type. Please refer to:
+	// https://github.com/vmware-tanzu/vm-operator/blob/main/api/v1alpha1/virtualmachine_types.go#L149.
+	NSXTVPCSubnetSetNetworkType = "nsx-t-subnetset"
+	// NSXTVPCSubnetSetSelectorKey has to be aligned with VM Operator design.
```
These are the annotation keys designed for the VirtualMachineService and LoadBalancer type service, originally defined in CAPW. It's quite uncertain if vm-operator/ncp/nsx-operator utilize them anymore. We should contemplate removing all of them once their usage in vm-operator/ncp/nsx-operator has been confirmed and the adjustment could be undertaken in a separate, dedicated fix.

chrischdi replied:
Yeah, and if we are not sure and don't get answers: maybe not introduce it and test if it works without having it.

silvery1622 replied:
Yes, the testing could proceed without the annotation. However, I suggest that we should still consult with the vmoperator team to guarantee that we haven't overlooked any corner cases in our tests.

silvery1622 commented:
A big thanks to @sbueringer and @chrischdi for the awesome reviews and input. I've managed to deal with most of what you pointed out. The one thing I'm still sorting out is the annotation, but I'm about to kick off some tests without it. Could our goal still be to include this in the v1.10.0 release?

sbueringer commented:
> The one thing I'm still sorting out is the annotation, but I'm about to kick off some tests without it. Could our goal still be to include this in the v1.10.0 release?

Yup, no worries

sbueringer reviewed (commented):
I think some of the comments or changes got lost (or not yet pushed). PTAL at the open conversations above

(Please note the "4 hidden conversations" (by GitHub))

silvery1622 commented on pkg/services/network/nsxt_vpc_provider.go:
```diff
@@ -0,0 +1,196 @@
+	}
+
+	if !hasReadyCondition {
+		return errors.Errorf("The ready status for subnetset in cluster %s has not been set", types.NamespacedName{Namespace: namespace, Name: clusterName})
```
Sure

silvery1622 commented:
I've made some changes based on feedback from the review:

Removed NSXTVPCSelectorKey and let GetVMServiceAnnotations returns a blank map. Also updated the UT for it.
Sorted out a situation where there's no 'ready' condition in subnetset.Status when we're running verifyNsxtVpcSubnetSetStatus. I've also added a UT case for it.
Used klog for logging in ProvisionClusterNetwork and included the network namespace and name in log.
Created a mock up of 'ProvisionClusterNetwork' to sidestep having to check the subnetset ready condition. There are separate cases to deal with when the subnet ready condition hasn't been set or when it's false.
Got rid of the double-checking for errors and switched the name from 'localerr' to just 'err'.

chrischdi commented on pkg/services/network/nsxt_vpc_provider.go:
```diff
@@ -0,0 +1,186 @@
+	hasReadyCondition := false
+
+	for _, condition := range subnetset.Status.Conditions {
+		if condition.Type == nsxopv1.Ready {
+			hasReadyCondition = true
+			if condition.Status != corev1.ConditionTrue {
+				conditions.MarkFalse(vspherecluster, vmwarev1.ClusterNetworkReadyCondition, vmwarev1.ClusterNetworkProvisionFailedReason, clusterv1.ConditionSeverityWarning, condition.Message)
+				return errors.Errorf("subnetset ready status is: '%s' in cluster %s. reason: %s, message: %s",
+					condition.Status, types.NamespacedName{Namespace: namespace, Name: clusterName}, condition.Reason, condition.Message)
+			}
+		}
```
Makes it a bit more readable to me :-)

```suggestion
		if condition.Type != nsxopv1.Ready {
			continue
		}
		hasReadyCondition = true
		if condition.Status != corev1.ConditionTrue {
			conditions.MarkFalse(vspherecluster, vmwarev1.ClusterNetworkReadyCondition, vmwarev1.ClusterNetworkProvisionFailedReason, clusterv1.ConditionSeverityWarning, condition.Message)
			return errors.Errorf("subnetset ready status is: '%s' in cluster %s. reason: %s, message: %s",
				condition.Status, types.NamespacedName{Namespace: namespace, Name: clusterName}, condition.Reason, condition.Message)
		}
```

silvery1622 replied:
Sure. The use continue could make the control flow more clear.

chrischdi reviewed (commented):
one more readability nit

sbueringer commented:
/test ?

sbueringer commented:
Thank you very much!

/lgtm
/approve

Merge pending CI (want to make sure we don't break CI that close to the release)


/test pull-cluster-api-provider-vsphere-e2e-govmomi-conformance-ci-latest-main
/test pull-cluster-api-provider-vsphere-e2e-govmomi-conformance-main
/test pull-cluster-api-provider-vsphere-e2e-govmomi-main
/test pull-cluster-api-provider-vsphere-e2e-govmomi-upgrade-1-29-1-30-main
/test pull-cluster-api-provider-vsphere-e2e-supervisor-conformance-ci-latest-main
/test pull-cluster-api-provider-vsphere-e2e-supervisor-conformance-main
/test pull-cluster-api-provider-vsphere-e2e-supervisor-main
/test pull-cluster-api-provider-vsphere-e2e-supervisor-upgrade-1-29-1-30-main
/test pull-cluster-api-provider-vsphere-e2e-vcsim-govmomi-main
/test pull-cluster-api-provider-vsphere-e2e-vcsim-supervisor-main

silvery1622 commented:
Updated downstream test log with latest code.
[capv-controller-manager-1.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14916792/capv-controller-manager-1.log)
[capv-controller-manager-2.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14916795/capv-controller-manager-2.log)
[cluster.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14916796/cluster.log)
[virtualmachine.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14916797/virtualmachine.log)
[virtualmachineservice.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14916798/virtualmachineservice.log)
[vspherecluster.log](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/files/14916799/vspherecluster.log)
